src/mds/ScatterLock.h

```diff
@@ -56,7 +56,7 @@ class ScatterLock : public SimpleLock {
   void set_xlock_snap_sync(MDSContext *c)
   {
     ceph_assert(get_type() == CEPH_LOCK_IFILE);
-    ceph_assert(state == LOCK_XLOCK || state == LOCK_XLOCKDONE);
+    ceph_assert(state == LOCK_XLOCK || state == LOCK_XLOCKDONE || state == LOCK_LOCK_XLOCK);
     state = LOCK_XLOCKSNAP;
```

Comment:
I'm not sure this is a valid state transition from `LOCK_LOCK_XLOCK`. Would a wait for the journal entry to complete be more appropriate?

Reply:
> I'm not sure this is a valid state transition from LOCK_LOCK_XLOCK.

Yeah, checked the code more carefully, it seems this is invalid state transition.

>  Would a wait for the journal entry to complete be more appropriate?

This is in one of the journal entry's `complete()`. After another journal entry's `complete()` the filelock's state will be in  `LOCK_LOCK` or some other states, which will also be a invalid state transition to XLOCKSNAP.

I have updated it by suspending the lock state to `LOCK_XLOCKDONE` if inode truncating is needed.